### ISSUE TITLE
Use of int8 for variant_contig results in integer overflow with fragmented reference genomes

### ISSUE THREAD

**Issue by timothymillar** [pystatgen/sgkit#584]
Many (non human) reference genomes contain 1000s of contigs that have not been assembled into full chromosomes.
Currently the `variant_contig` array  is hard coded as int8 ([line](https://github.com/pystatgen/sgkit/blob/main/sgkit/io/vcf/vcf_reader.py#L267)) which results in integer overflow making it impossible to join variants to their contig.

**Comment by hammer** [comment]
We should probably revisit how we're parsing and representing contig metadata more generally, as noted in https://github.com/pystatgen/sgkit/issues/464 as well.

@timothymillar do you have any example VCFs to share that overflow the `int8` limit? It's trivial to manually construct such a VCF but I generally prefer to have real world data in our tests.

**Comment by timothymillar** [comment]
@hammer agreed, real data always hits more edge cases. I'll look into an example VCF for this. 

For reference I ran into this issue when using the [Red5](https://plants.ensembl.org/Actinidia_chinensis/Info/Index) kiwifruit genome.

**Comment by tomwhite** [comment]
We use int16 for bgen and PLINK, so we should probably just change VCF to be the same.

**Comment by jeromekelleher** [comment]
We should make it an option to specify the dtype for ``variant_contig`` probably - even ``int16`` will overflow sometimes. There are lots of VCFs out there with huge numbers of contigs.

Although, I guess this is the sort of thing we should be able to query the IO library for ("how many contigs are there" should be efficiently computable on any indexed VCF), so we should be able to automatically detect the minimal dtype. Even then though, I suppose people might want to manually specify the dtype, for their own reasons.

**Comment by alxsimon** [comment]
Is there a workaround at the moment? Working with a non model species with > 50000 contigs

**Comment by jeromekelleher** [comment]
Nice - great to see you pushing the limits here @alxsimon! @tomwhite any thoughts on how we should address this?

**Comment by tomwhite** [comment]
I'm afraid I can't think of a workaround for this. The fix that @jeromekelleher sketched out above should be fairly straightforward though, so I'll work on a PR to fix it.

**Comment by alxsimon** [comment]
Thanks @tomwhite and @jeromekelleher, I'll wait for the upstream fix.

**Comment by alxsimon** [comment]
Quick and dirty workaround to reimport the contig names from the vcf, in case someone else is looking for a way to do this in the meantime.
```
from cyvcf2 import VCF

variant_contig_name = np.empty(ds.dims['variants'], dtype="O")
variant_position = np.empty(ds.dims['variants'], dtype="i4")
for idx, variant in enumerate(VCF(bcf_file)):
    variant_contig_name[idx] = variant.CHROM
    variant_position[idx] = variant.POS

ds = ds.merge(xr.DataArray(variant_contig_name, coords=ds.coords, dims=['variants'], name='variant_contig_name'))
```

**Comment by tomwhite** [comment]
Nice!

I've created a fix in #667. Hopefully we can get that merged soon for you to use.
